src/cognito/auth.ts

```diff
@@ -0,0 +1,185 @@
+import { AuthenticationDetails, CognitoUser, CognitoUserPool } from 'amazon-cognito-identity-js';
+import { getEnv } from '../utils/common';
+import { isBeta } from '../utils/common';
+
+async function fetchData() {
+  try {
+    const response = await fetch(`${isBeta() ? '/beta' : ''}/apps/chrome/env.json`);
+    const jsonData = await response.json();
+    return jsonData;
+  } catch (error) {
+    console.log(error);
+  }
+}
+
+function getSearchParams(url: string): { [key: string]: string } {
+  const searchString = new URL(url).search;
+  const searchParams = new URLSearchParams(searchString);
+  const searchParamsObject: { [key: string]: string } = {};
+  for (const [key, value] of searchParams) {
+    searchParamsObject[key] = value;
+  }
+  return searchParamsObject;
+}
+
+export async function getTokenWithAuthorizationCode() {
+  const code = getSearchParams(window.location.href).code;
+  const refreshToken = localStorage.getItem('REFRESH_TOKEN');
+
+  const data = await fetchData();
+  const dataConfig = getEnv() === 'gov' ? data.poolData?.prod : getEnv() === 'govStage' ? data.poolData?.stage : null;
+  const loginUrl = `${dataConfig?.ssoUrl}/login?client_id=${dataConfig?.ClientId}&response_type=code&scope=openid&redirect_uri=${dataConfig?.redirectUri}`;
+
+  const redirectUri = dataConfig?.redirectUri;
+  if (!code && !refreshToken) {
+    window.location.href = loginUrl;
+  }
+  if (refreshToken) {
+    try {
+      const response = await fetch(`${dataConfig?.ssoUrl}/oauth2/token`, {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded',
+        },
+        body: `grant_type=refresh_token&client_id=${dataConfig?.ClientId}&refresh_token=${localStorage.getItem('REFRESH_TOKEN')}`,
+      });
+      if (!response.ok) {
+        window.location.href = loginUrl;
+        throw new Error(`Request failed with status code ${response.status}`);
+      }
+
+      const tokens = await response.json();
+
+      localStorage.setItem('ACCESS_TOKEN', tokens.access_token);
+      return tokens.access_token;
+    } catch (error) {
+      console.error(error);
+    }
+  }
+  try {
+    const response = await fetch(`${dataConfig?.ssoUrl}/oauth2/token`, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+      body: `grant_type=authorization_code&client_id=${dataConfig?.ClientId}&redirect_uri=${redirectUri}&code=${code}`,
+    });
+
+    if (!response.ok) {
+      throw new Error(`Request failed with status code ${response.status}`);
+    }
+
+    const tokens = await response.json();
+    localStorage.setItem('REFRESH_TOKEN', tokens.refresh_token);
+    localStorage.setItem('ACCESS_TOKEN', tokens.access_token);
```

Comment:
We should probably store the storage keys in some constants so we have them in one place if we have to change them later on.